quickwit-search/src/root.rs

```diff
@@ -209,13 +228,79 @@ pub async fn root_search(
         &assigned_leaf_search_jobs,
         search_request_with_offset_0.clone(),
     )
     .await?;
 
-    analyze_errors(&result_per_node_addr);
+    let analyze_result = analyze_errors(&result_per_node_addr);
+    if let Some(retry_action) = analyze_result.as_ref() {
+        // Create a job for fetching docs and assign the splits that the node is responsible for based on the job.
+        //
+        let leaf_search_jobs: Vec<Job> = job_for_split(
+            &retry_action.retry_split_ids.iter().collect(),
+            &split_metadata_map,
+        );
+
+        let retry_assigned_leaf_search_jobs = client_pool
+            .assign_jobs(leaf_search_jobs, &Some(retry_action.nodes_to_avoid.clone()))
+            .await?;
+        // Perform the query phase.
+        let result_per_node_addr_new = execute_search(
+            &retry_assigned_leaf_search_jobs,
+            search_request_with_offset_0.clone(),
+        )
+        .await?;
+
+        // clean out old errors
+        let complete_errors = result_per_node_addr
+            .iter()
+            .filter(|(_addr, res)| res.is_err())
+            .map(|(addr, _err)| *addr)
+            .collect_vec();
+        for err_addr in complete_errors {
+            result_per_node_addr.remove(&err_addr);
+        }
+        // remove partial, retryable errors
+        for result in result_per_node_addr.values_mut().flatten() {
+            let failed_splits: Vec<SplitSearchError> = std::mem::take(&mut vec![]);
+            result.failed_splits = failed_splits
+                .into_iter()
+                .filter(|err| !err.retryable_error)
+                .collect_vec();
+        }
+
+        for (addr, new_result) in result_per_node_addr_new {
+            if let Some(Ok(orig_res)) = result_per_node_addr.get_mut(&addr) {
+                // old one exists and is_ok and new one is_ok
+                if let Ok(result) = new_result {
+                    orig_res.num_hits += result.num_hits;
+                    orig_res.num_attempted_splits += result.num_attempted_splits;
+                    orig_res
+                        .failed_splits
+                        .extend(result.failed_splits.into_iter());
+                    orig_res
+                        .partial_hits
+                        .extend(result.partial_hits.into_iter());
+                } else if let Err(err) = new_result {
+                    // old one exists and is_ok and new one is not ok
+                    orig_res
+                        .failed_splits
+                        .extend(err.split_ids.iter().map(|split_id| SplitSearchError {
+                            error: err.search_error.to_string(),
+                            split_id: split_id.to_string(),
+                            retryable_error: true,
+                        }));
+                }
+            } else if let Some(Err(err)) = result_per_node_addr.get_mut(&addr) {
+                panic!("unexpected error leftover: {:?}", err);
```

Comment:
Wait... When does that happen?

Reply:
never, that's why it's a panic. We remove retryable errors before we retry and complete NodeSearchErrors are considered retryable currently